flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java

```diff
@@ -0,0 +1,243 @@
+     * <li>There exists one partition that contains more than one double array.
+     * <li>The length of the double array is not consistent among all partitions.
+     *
+     * @param input The input data stream.
+     * @return The result data stream.
+     */
+    public static DataStream<double[]> allReduceSum(DataStream<double[]> input) {
+        return AllReduceImpl.allReduceSum(input);
+    }
+
+    /**
+     * Collects distinct values in a bounded data stream. The parallelism of the output stream is 1.
+     *
+     * @param <T> The class type of the input data stream.
+     * @param input The bounded input data stream.
+     * @return The result data stream that contains all the distinct values.
+     */
+    public static <T> DataStream<T> distinct(DataStream<T> input) {
+        return input.transform(
+                        "distinctInEachPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(input.getParallelism())
+                .transform(
+                        "distinctInFinalPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(1);
+    }
+
+    /**
+     * Applies a {@link MapPartitionFunction} on a bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param func The user defined mapPartition function.
+     * @param <IN> The class type of the input element.
+     * @param <OUT> The class type of output element.
+     * @return The result data stream.
+     */
+    public static <IN, OUT> DataStream<OUT> mapPartition(
+            DataStream<IN> input, MapPartitionFunction<IN, OUT> func) {
+        TypeInformation<OUT> resultType =
+                TypeExtractor.getMapPartitionReturnTypes(func, input.getType(), null, true);
+        return input.transform("mapPartition", resultType, new MapPartitionOperator<>(func))
+                .setParallelism(input.getParallelism());
+    }
+
+    /**
+     * Sorts the elements in each partition of the input bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param comparator The comparator used to sort the elements.
+     * @param <IN> The class type of input element.
+     * @return The sorted data stream.
+     */
+    public static <IN> DataStream<IN> sortPartition(
+            DataStream<IN> input, TypeComparator<IN> comparator) {
+        return input.transform(
+                        "sortPartition", input.getType(), new SortPartitionOperator<>(comparator))
+                .setParallelism(input.getParallelism());
+    }
+
+    /**
+     * A stream operator to compute the distinct values in each partition of the input bounded data
+     * stream.
+     */
+    private static class DistinctPartitionOperator<T> extends AbstractStreamOperator<T>
+            implements OneInputStreamOperator<T, T>, BoundedOneInput {
```

Comment:
I have removed `DistinctOperator`.

BTW, I am not sure whether `AbstractStreamOperator::close` can replace `BoundedOneInput`. If so, why do Flink provides a separate interface for `BoundedOneInput`?
if `AbstractStreamOperator::close`

Reply:
I think `BoundedOneInput ` could still be useful when the operator does not extend `AbstractStreamOperator`. There may be other reasons.

Reply:
To make a fair comparison, I copy-paste the java doc as follows:

- StreamOperator#close: `This method is called at the very end of the operator's life, both in the case of a successful completion of the operation, and in the case of a failure and canceling.`
- BoundedOneInput#endInput: `It is notified that no more data will arrive on the input.`

I think the main difference here is that whether one data stream is bouned or unbounded. Since (1) `StreamOperator#close` is called no matter the input is bounded or unbounded and (2) `BoundedOneInput#endInput` is called at the end of a bounded input.

As we can only call mapPartition on a bounded data stream, I think `BoundedOneInput#endInput` is a better choice.

Reply:
Hmm..  do you mean that we can not use `StreamOperator#close` here? Can you explain why?

If we can use both `StreamOperator#close` and `BoundedOneInput#endInput`, why is `BoundedOneInput#endInput` a better choice?

Reply:
Hmm, I am not sure `StreamOperator#close` works or not here.  My point is that `mapPartition` should only be applied on a bounded data stream.
According to the java doc above, `BoundedOneInputp#endInput` is a better choice from semantics.

What about we leave this question here and see how does @gaoyunhaii  think?

Reply:
Hi, I could not find the class in the PR, currently endInput is called only on normal finish (namely all records are processed), and close is called on both normal finish and exceptional exit (cancel / failover), thus I think logically we should put business in endInput and resource cleanup in close.